src/core/sr.ml

```diff
@@ -102,50 +87,109 @@ let check_rule : sym StrMap.t -> sym * pp_hint * rule Pos.loc -> unit =
     | Wild        -> assert false (* Cannot appear in LHS. *)
     | TRef(_)     -> assert false (* Cannot appear in LHS. *)
   in
-  let lhs = List.map (fun p -> Bindlib.unbox (to_m 0 p)) r.elt.lhs in
-  let lhs = Basics.add_args (Symb(s,h)) lhs in
-  let metas = Array.map (function Some m -> m | None -> assert false) metas in
-  (* We substitute the RHS with the corresponding metavariables. *)
-  let fn m =
-    let xs = Array.init m.meta_arity (Printf.sprintf "x%i") in
-    let xs = Bindlib.new_mvar mkfree xs in
-    let e = Array.map _Vari xs in
-    TE_Some(Bindlib.unbox (Bindlib.bind_mvar xs (_Meta m e)))
+  let lhs_args = List.map (fun p -> Bindlib.unbox (to_m 0 p)) rule.lhs in
+  let lhs = Basics.add_args (Symb(s,h)) lhs_args in
+  let lhs_metas =
+    Array.map (function Some m -> m | None -> assert false) lhs_metas
   in
-  let te_envs = Array.map fn metas in
-  let rhs = Bindlib.msubst r.elt.rhs te_envs in
-  (* Infer the type of the LHS and the constraints. *)
+
+  (* Build a map LHS pattern variable -> term_env variable. *)
+  let sym_name m = "&" ^ match m.meta_name with
+                         | Some(n) -> n
+                         | None    -> string_of_int m.meta_key
+  in
+  let map =
+    let map = ref StrMap.empty in
+    let fn i m =
+      map := StrMap.add (sym_name m) (rule.vars.(i), m.meta_arity) !map
+    in
+    Array.iteri fn lhs_metas;
+    !map
```

Comment:
I would probably do:
```ocaml
let map = ref StrMap.empty in
let fn x m = map := StrMap.add (sym_name m) (x, m.meta_arity) !map in
Array.iter2 fn rule.vars lhs_metas; !map
```
or
```ocaml
let fn (i, map) m =
  let elt = (rule.vars.(i), m.meta_arity) in
  (i + 1,  StrMap.add (sym_name m) elt map)
in
snd (Array.fold_left fn (0, StrMap.empty) lhs_metas)
```
or I would define an `Array.foldi` in `Extra` and do:
```ocaml
let fn map i m =
  StrMap.add (sym_name m) (rule.vars.(i), m.meta_arity) map
in
Array.foldi fn StrMap.empty lhs_metas
```